fix!(codecs): Use '\0' delimiter as default stream decode framer

jorgehermo9 opened the pull request:
Closes #20768.

I think that in my local environment some test fail but has nothing to do with the changes I did. Could you trigger them here and if its fails too in the CI I review what could happen?

Thanks

jorgehermo9 commented on lib/codecs/src/decoding/framing/character_delimited.rs:
```diff
@@ -53,7 +59,7 @@ pub struct CharacterDelimitedDecoderOptions {
 impl CharacterDelimitedDecoderOptions {
     /// Create a `CharacterDelimitedDecoderOptions` with a delimiter and optional max_length.
-    pub fn new(delimiter: u8, max_length: Option<usize>) -> Self {
+    pub const fn new(delimiter: u8, max_length: Option<usize>) -> Self {
```
Noticed this and didn't hurt

jorgehermo9 commented on lib/codecs/src/encoding/mod.rs:
```diff
@@ -362,12 +362,14 @@ impl SerializerConfig {
             | SerializerConfig::NativeJson
             | SerializerConfig::RawMessage
             | SerializerConfig::Text(_) => FramingConfig::NewlineDelimited,
+            SerializerConfig::Gelf => {
```
In order to be consistent with the `default_stream_framing` of the `DeserializerConfig`. Noticed that this change was missing in #18008. Changed it here, but tell me if you want this PR to not do anything out of scope from #20768.

jszwedko replied:
Huh, yeah, this does seem like an oversight. I'm trying to understand the impact of it though since the original PR did seem to resolve the issue for people 🤔

jorgehermo9 replied:
It seems that the changed struct was `EncodingConfigWithFraming` (#18419).

It seems that a few places relies on `EncodingConfigWithFraming` to get this default behaviour, and other places uses this method `SerializerConfig::default_stream_framing`. For example, see this line https://github.com/vectordotdev/vector/blob/67a5e4682138721575b22369cf5ba629cd86ce55/src/components/validation/resources/mod.rs#L113
Where we can see its relying on `SerializerConfig::default_stream_framing` in order to get the default framing for a given encoder.

There are a lot of sinks that relies on `EncodingConfigWithFraming` having a Option<FramingConfig> in order to calculate that default (https://github.com/search?q=repo%3Avectordotdev%2Fvector%20EncodingConfigWithFraming&type=code) https://github.com/vectordotdev/vector/blob/67a5e4682138721575b22369cf5ba629cd86ce55/src/codecs/encoding/config.rs#L110

jszwedko replied:
Aha, gotcha, thanks for looking into that.

jorgehermo9 commented on src/codecs/decoding/decoder.rs:
```diff
@@ -138,4 +138,33 @@ mod tests {
     }
+
+    #[tokio::test]
+    async fn gelf_streaming_decoding() {
```
Added this test, I don't know if its correct to place it here. Didn't find anywhere else to place it.

I'm new in this project and can't find some modules easily. 
Maybe this test is more of an integration test, but couldn't find any related in the root `/tests` directory. I'm missing something?

It would be nice to have something like a TCP socket integration test where we test this same behavior...As the `default_stream_framing` is also called in https://github.com/vectordotdev/vector/blob/e982f6679e9d1526d856efa462e728774b52cf34/src/sources/socket/mod.rs#L123.

I thought of adding a similar test like this in https://github.com/vectordotdev/vector/blob/e982f6679e9d1526d856efa462e728774b52cf34/src/sources/socket/mod.rs#L391, what do you think?
In that case, I would create a TCP config setting only `decoding = DeserializerConfig::from(GelfDeserializerConfig)`(and leaving the framing as None in that config) and see how the TCP initialization takes the default framer.

What are your thoughts of all of this?

jorgehermo9 commented:
Please, note that this is a breaking change as we are changing the default framing. 

Maybe someone configured a source to send '\n' delimited gelf messages and left the vector config as the default, asumming that it would be correct.

In my opinion, we should adhere to the stardard and go ahead with this "breaking change"

jszwedko reviewed (commented):
Thanks for this fix @jorgehermo9!

Do you mind adding a changelog fragment? See https://github.com/vectordotdev/vector/blob/master/changelog.d/README.md

I think the test you added is ok. I'd actually be content with just testing the the default is set correctly without testing that it actually decode with the default config and `\0` delimited input since the character delimited codec is tested separately. If anything, you could add a test to `lib/codecs/src/decoding/framing/character_delimited.rs` that tests `\0` works as a delimiter, but I don't think it is strictly necessary.

jorgehermo9 commented:
Thanks for the review! I'll submit the changes asap.

> I'd actually be content with just testing the the default is set correctly without testing that it actually decode with the default config and \0 delimited input since the character delimited codec is tested separately

Right, I will narrow the scope of that test, no problem. It seemed too simple for me to just test that, but it doesn't have sense to test the \0 framing there...

jorgehermo9 commented on lib/codecs/src/decoding/mod.rs:
```diff
@@ -334,14 +334,16 @@ impl DeserializerConfig {
             DeserializerConfig::Syslog(_) => FramingConfig::NewlineDelimited(Default::default()),
             DeserializerConfig::Vrl(_) => FramingConfig::Bytes,
+            DeserializerConfig::Gelf(_) => {
+                FramingConfig::CharacterDelimited(CharacterDelimitedDecoderConfig::new(0))
```
I was thinking about this these days... I think that GELF Http maybe uses newline delimiter? I can't find anything related to streaming gelf in Http here https://go2docs.graylog.org/5-0/getting_in_log_data/gelf.html, but maybe someone will rely on that for http sources also and this will be a breaking change in those cases too, not only with TCP.

If you are ok with that, I will include that breaking changes and the "fix" (manually setting the newline delimiter in `framing.method`) in the changelog fragment

jszwedko replied:
Interesting, I haven't heard of anyone using GELF over HTTP, but it seems plausible. TCP seems to be the dominant transport so I think it makes sense to configure the defaults to work well with that.

jorgehermo9 commented:
Hi @jszwedko, I'm sorry for addressing the changes that late. I simplified that test and moved it to another file.

Could you trigger the CI and see if everything works ok?

jszwedko commented on changelog.d/20768-default_gelf_stream_framing.breaking.md:
```diff
@@ -0,0 +1,31 @@
+Now the GELF codec with stream-based sources uses null byte (`\0`) by default as messages delimiter instead of newline (`\n`) character.
```
```suggestion
Now the GELF codec with stream-based sources uses null byte (`\0`) by default as messages delimiter instead of newline (`\n`) character. This better matches GELF server behavior.
```

jszwedko commented on changelog.d/20768-default_gelf_stream_framing.breaking.md:
```diff
@@ -0,0 +1,31 @@
+
+In order to maintain the previous behavior, you must set the `framing.method` option to the `character_delimited` method and the `framing.character_delimited.delimiter` option to `\n` when using GELF codec with stream-based sources.
+
+### Example configuration change for socket source
```
Awesome, thanks for including the example for using the previous default.

Small nit that we try to use YAML rather than TOML as the default these days.

jorgehermo9 replied:
Done!

jszwedko reviewed (approved):
Awesome, thanks again for this change @jorgehermo9 !

jszwedko reviewed (approved):
Thanks again for this!

jorgehermo9 commented:
Thank you very much for your prompt review!
